lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -661,74 +728,411 @@ namespace Js
+        // However, passing the locale through uloc_forLanguageTag -> uloc_toLanguageTag gets us most of the way there
+        // by replacing some(?) values, correctly capitalizing the tag, and re-ordering extensions
+        int parsedLength = 0;
+        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int forLangTagResultLength = uloc_forLanguageTag(langtag8, localeID, ULOC_FULLNAME_CAPACITY, &parsedLength, &status);
+        if (status == U_ILLEGAL_ARGUMENT_ERROR)
+        {
+            // The string passed in to NormalizeLanguageTag has already passed IsStructurallyValidLanguageTag.
+            // However, duplicate unicode extension keys, such as "de-u-co-phonebk-co-phonebk", are structurally
+            // valid according to RFC5646 yet still trigger U_ILLEGAL_ARGUMENT_ERROR
+            // V8 ~6.2 says that the above language tag is invalid, while SpiderMonkey ~58 handles it.
+            // Until we have a more spec-compliant implementation of CanonicalizeLanguageTag, err on the side
+            // of caution and say it is invalid.
+            // NOTE: make sure we check for `undefined` at the platform.normalizeLanguageTag callsite.
             return scriptContext->GetLibrary()->GetUndefined();
         }
 
-        JavascriptString *argString = JavascriptString::FromVar(args.Values[1]);
+        // forLangTagResultLength can be 0 if langtag is "und".
+        // uloc_toLanguageTag("") returns "und", so this works out (forLanguageTag can return >= 0 but toLanguageTag must return > 0)
+        ICU_ASSERT(status, forLangTagResultLength >= 0 && ((charcount_t) parsedLength) == langtag->GetLength());
 
-        HRESULT hr;
-        JavascriptString *retVal;
+        char canonicalized[ULOC_FULLNAME_CAPACITY] = { 0 };
+        int toLangTagResultLength = uloc_toLanguageTag(localeID, canonicalized, ULOC_FULLNAME_CAPACITY, true, &status);
+        ICU_ASSERT(status, toLangTagResultLength > 0);
 
-#if defined(INTL_ICU)
-        // `normalized` will be filled by converting a char* (of utf8 bytes) to char16*
-        // Since `len(utf8bytes) >= len(to_char16s(utf8bytes))`,
-        // Therefore the max length of that char* (ULOC_FULLNAME_CAPACITY) is big enough to hold the result (including null terminator)
-        char16 normalized[ULOC_FULLNAME_CAPACITY] = { 0 };
-        size_t normalizedLength = 0;
-        hr = NormalizeLanguageTag(argString->GetSz(), argString->GetLength(), normalized, &normalizedLength);
-        if (FAILED(hr))
+        // allocate toLangTagResultLength + 1 to leave room for null terminator
+        char16 *canonicalized16 = RecyclerNewArrayLeaf(scriptContext->GetRecycler(), char16, toLangTagResultLength + 1);
+        size_t canonicalized16Len = 0;
+        HRESULT hr = utf8::NarrowStringToWideNoAlloc(
+            canonicalized,
+            toLangTagResultLength,
+            canonicalized16,
+            toLangTagResultLength + 1,
+            &canonicalized16Len
```

Comment:
I changed the type of this in a recent PR, this is now a `charcount_t*` parameter